### PR TITLE
Initial draft of k8s without helm reference

### PR DESCRIPTION
**Pull request by IAmHughes**
This PR adds my initial draft of the deploy guide/reference for setting up Sourcegraph on Kubernetes (k8s) without using Helm.

- **Problem Statement**: Currently the Sourcegraph documentation for deploying can have a number of splitting paths making it hard to navigate
- **Proposed Solution**: By creating references like this, the CE/CTE team can hand out a reference PDF (converted from this source controlled markdown file) which assist Sourcegraph Admins in navigating through our documentation in order to properly create their clusters and deploy Sourcegraph to them.
  - **Why this?**: This is to serve as a specific use-case reference (deploying without Helm to k8s) and is not a replacement for official documentation, but a map to the documentation for this use case. This balances the technical debt involved with maintaining detailed documentation and having two sources of truth with a short, easily digestible map to the source of truth (official documentation).

@sourcegraph/customer-training-engineering I would love your 👀  and a 👍  or 💬  letting me know your thoughts 💭. I'm very open to iterating on these in the future, but wanted to get something pushed for a template to use on additional deployment methods.

### CODE REVIEW

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [inline collateral/deployment/k8s-without-helm-reference.md:4]:
```suggestion
This material is intended to be used as a handout to customers working with a Sourcegraph Customer Engineer to deploy their own instance. The materials are meant to serve as an overview and resource to aid in deep diving into Sourcegraph's documentation and common deployment tasks.
```

Clarifying here to add SG explicitly + confirm the task at hand, since we do lots of stuff

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [inline collateral/deployment/k8s-without-helm-reference.md:19]: Up to you on whether you want to leave this, but this isn't required to deploy, and suggesting that it is confuses a lot of people historically (typically they receive the license after the instance is stood up.)

```suggestion
```

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [inline collateral/deployment/k8s-without-helm-reference.md:22]:
```suggestion
  - Support for Persistent Volumes (SSDs strongly recommended)
```
I might even change this to "required", the delivery team treats them as a prereq.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [inline collateral/deployment/k8s-without-helm-reference.md:23]: I defer to you on how you want to phrase this, but the resource estimator often underestimates resources by about 40%, so this functionally is really "use the resource estimator as a starting-off point but pick values in conjunction with the CE"

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [inline collateral/deployment/k8s-without-helm-reference.md:30]: @mike-r-mclaughlin Are the values suggested in this doc good or no?

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [inline collateral/deployment/k8s-without-helm-reference.md:39]:
```suggestion
 ```
Same reason as above.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [inline collateral/deployment/k8s-without-helm-reference.md:34]: I worry that this is just pointing them to a very long doc without much guidance.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [inline collateral/deployment/k8s-without-helm-reference.md:53]:
```suggestion
Congratulations, Sourcegraph is now deployed and you can begin to configure your instance. We recommend you review the [Configurating Sourcegraph](https://docs.sourcegraph.com/admin/config) documentation next to [add a code host](https://docs.sourcegraph.com/admin/external_service), [configure authN and authZ](https://docs.sourcegraph.com/admin/auth), and more.
```
May want to consider linking to https://docs.sourcegraph.com/admin/config/authorization_and_authentication instead, since it's oriented more towards "if you want x do y" than the main auth docs.

**Review by emchap — COMMENTED** [review]
I would suggest that we link to information about backup, which I didn't see included: https://docs.sourcegraph.com/admin/deploy/migrate-backup#migration-and-backup-options

Are all the docs referenced in https://docs.google.com/document/d/1Cg4ttBfu35qVO9TT1AhDuKAu899xwR2-GpiVDGakOzE/edit (the predecessor to this document) referenced here? I think we'll have a hard time deprecating that document if not. 

I would love to get CE eyes on this as well— @christinelovett might be good to check with, since the doc I linked to is hers, originally. I just want to make sure that this meets the same needs so we can deprecate the other.

**Comment by christinelovett** [comment]
Thanks @emchap and nice work both. @IAmHughes lmk if you have any questions after reading thru the deployment template Emily linked

File: collateral/deployment/k8s-without-helm-reference.md
Comment (DaedalusG) [reply collateral/deployment/k8s-without-helm-reference.md:22]: +1 to Emily here, solid state drives are typically created as [required](https://docs.sourcegraph.com/admin/deployment_best_practices#kubernetes)

**Comment by DaedalusG** [comment]
Hey @IAmHughes this looks pretty good to me, by chance has anyone given this doc a run-through to deploy a cluster? If not this might be a good exercise to test things out. 

We have a really explicit [procedural doc for GCP k8s deployment](https://handbook.sourcegraph.com/departments/support/process/enablement/gcp-sg-k8s/) internally for CS. It'd be a good exercise to have someone try and spin up a cluster using this document on GCP and see how difficult it is compared to a more codehost specific and explicit doc.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:19]: @emchap are they modifying their manifest and applying again afterward? I guess if I was going to do this, I'd create my cluster and deploy with all my configs in the manifest file then later changes/tweaks to my manifest could be applied later if needed, but a license to get sourcegraph started seemed like a "first deploy" thing versus a later modification.

Totally open to your thoughts since you've gone through implementation here and have the insight.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:23]: Hmm, this is a good point actually. I think adding something about the CE works for sure, but in general we should have resources that work. If the resource estimator isn't working (which I think I heard elsewhere it isn't always accurate too), that tool should be removed or improved, right? I think there's a team working to improve the estimator but I'm not sure what the stage is.

How does this sound as a "for now" thing and hopefully in the future we can rely more on the resource estimator, or remove it if it doesn't work properly.

```suggestion
- Determine resource requirements to properly size your instance: The [Resource Estimator](https://docs.sourcegraph.com/admin/deploy/resource_estimator) can give you a baseline, however work with your Sourcegraph Customer Engineer to ensure you are sized appropriately
```

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [reply collateral/deployment/k8s-without-helm-reference.md:19]: @IAmHughes The license is added in-app and is unrelated to the deployment; it's usually given to them once the deployment is stood up, and the CE will run through checks with them at this point.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:30]: If they aren't, let's PR the original doc so we have a good single source of truth. 🙏

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:39]: Going to wait on this based on my [question above about that response](https://github.com/sourcegraph/customer-training/pull/5#discussion_r883038670)

(more so a note to myself on why I left the convo open, haha)

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [reply collateral/deployment/k8s-without-helm-reference.md:23]: Don't disagree! But it is the way it currently is (and has been for at least the last 18 months; the whole thing is a giant 100mb application and probably will just get completely rewritten). I believe Delivery is working on improvements this quarter and you can check with them on status.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [reply collateral/deployment/k8s-without-helm-reference.md:30]: I think the issue is more that the appropriate nodes to use vary deployment-by-deployment (https://stackoverflow.com/c/sourcegraph/questions/187/188#188 has more context on AWS node choice), but I defer to Mike.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:34]: @emchap the goal of this being a reference to a single source of truth, which should be our documentation, tells me that if that doc isn't good enough the team in charge of it needs to improve that doc.

I don't disagree, there's a LOT of information on that page about editing your manifest, but I wholeheartedly think this should be a reference to official documentation and documentation should be updated by the team that owns that to ensure it's correct, accurate, and usable. A single source of truth that is our official documentation makes way more sense in terms of supporting customers and long-term scalability.

@amenne I'm curious about your thoughts on this since we chatted a bit about docs this week too and content in the docs versus navigation of the docs.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:53]: Before I accept this @emchap, can you add the suggestion with the other auth piece? That makes sense to me if that's used and referenced more often than the main auth documentation. I think you're saying to do it this way?

```suggestion
Congratulations, Sourcegraph is now deployed and you can begin to configure your instance. We recommend you review the [Configurating Sourcegraph](https://docs.sourcegraph.com/admin/config) documentation next to [add a code host](https://docs.sourcegraph.com/admin/external_service), [configure authN and authZ](https://docs.sourcegraph.com/admin/config/authorization_and_authentication), and more.
```
The original link has a lot of auth information though on it, is it really not useful and shouldn't be here? Maybe both?

**Comment by IAmHughes** [comment]
> I would suggest that we link to information about backup, which I didn't see included: https://docs.sourcegraph.com/admin/deploy/migrate-backup#migration-and-backup-options

I didn't see where in that doc it's talked about setting something up during deploy/install that would enable backups. Is that somewhere else or something needed to setup during the deploy/install phase? If so, yes that should totally be in this reference. Otherwise, I think this reference being to deploy/install on k8s specifically, maybe we could add a link about backups to the end section where we threw in other configuration things like SSO/Auth, what do you think @emchap ?

> Are all the docs referenced in https://docs.google.com/document/d/1Cg4ttBfu35qVO9TT1AhDuKAu899xwR2-GpiVDGakOzE/edit (the predecessor to this document) referenced here? I think we'll have a hard time deprecating that document if not.

I'm not sure. That doc is great, and looks like something that would be perfect for a deployment guide that lives in the docs. @jwayman what do you think about a way to integrate guides like this into the docs, because this kind of step by step stuff is VERY valuable. This reference shouldn't get into details step by step for it's purpose, but this kind of step by step thing is exactly what I'd expect to find in a "guides" section of official documentation for any product I was wanting to use.

> I would love to get CE eyes on this as well— @christinelovett might be good to check with, since the doc I linked to is hers, originally. I just want to make sure that this meets the same needs so we can deprecate the other.

Agreed I'd love @sourcegraph/customer-engineering input in general 👍

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:19]: Interesting. @emchap would you say the best practice is to deploy my instance....then apply the manifest again with the license key later? As a best practice, it feels like having my license key initially when I modify my manifest for storage classes, ingress, etc makes more sense than applying it afterward.

I know you're speaking to how it has been, but is that really a best practice in general for standing up software like this? I haven't seen it that way in my past experience, but my past experience is a narrow sliver of the entire software ecosystem of course. Is there a benefit to adding the license later?

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:23]: > Don't disagree! But it is the way it currently is (and has been for at least the last 18 months; the whole thing is a giant 100mb application and probably will just get completely rewritten). I believe Delivery is working on improvements this quarter and you can check with them on status.

@amenne tagging you just for awareness, @jwayman as well. This tool is in the docs and isn't useful in it's current state, to the extent that we have been calling out that working with a CE is going to be a better route. I'm going to keep the link to the. resource estimator and the suggested change above of adding in CE language, but I wanted to make sure this is elevated again that we're saying here that our official docs aren't sufficient without CE involvement.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:30]: Will totes wait for @mike-r-mclaughlin, thanks!

I do think regardless if the information isn't correct in our official documentation we need a better way to show that in the documentation. This is to reference the official docs, not be a separate documentation piece (or a second source of truth).

File: collateral/deployment/k8s-without-helm-reference.md
Comment (IAmHughes) [reply collateral/deployment/k8s-without-helm-reference.md:53]: @emchap that looks great to me!

**Comment by jwayman** [comment]
I'll preface this by saying it's a very long response, but some of this probably needs to be captured more formally. I also think a sync conversation with the right parties in attendance might be better. However, I've attempted to provide as much detail as possible until we can coordinate.

So I think some context might explain some of the questions:

1. First, there's really been no one (at least to my knowledge) managing docs from a curation and custodial perspective for a number of years. I started at the end of January, and the person on the team (@JeanPaulSG started two weeks ago). The problem(s) described here are present in almost any part of the Admin/Deployment docs, including across all deployment types and various cloud providers/platforms. I'm certain there is content far and wide that we need to consolidate. This will take some time to unravel and I'm thankful for deep insight like what you've provided here - so thank you!
2. Next, when I arrived the directive was that we didn't want anyone self-servicing K8s on their own given the complexities associated with Kustomize and the trouble we had seen for customers. This level of rhetoric is present in the new deployment overview I created, as well as the K8s docs themselves, and has seemed to be echoed pretty congruently across most areas of our business. However, to address a good portion of those complexities and give customers the best possible focus, we invested in the K8s with Helm templates (which were just released at the end of the last quarter). I don't know if that's changed or been modified, but I think we should get some clarity on what we are recommending to customers and what we're focusing our time on. (cc @briemcnally @ryphil and @anna-mmikhova).
3. The docs should absolutely be "**a** source of truth", but they likely will not have every potential deployment path and option deeply documented from end-to-end. It should be "**the** source of truth" for the "**happy path**" where users that intend to self-service can follow our docs and achieve a successful deployment. I would also go as far as it should be "**the** source of truth" for reference information for all deployment types and supported Setup/Admin paths. I''m being very deliberate with the articles and terminology there. I fully expect certain scenarios will still have instructions that exist in a Google Doc that provide specific, non-self-service guidance to help organizations with particular deployment (not sure what that is, but it seems K8s via Kustomize might be heading that way given the release of K8s via Helm).
4. As reference material, and I am specifically citing [the definition we intend to move forward with](https://documentation.divio.com/reference/), the docs will continue to be delivered and maintained by the entire Product organization (as they have been) with custodial duties and guidance with regard to things like information architecture and templates (to guide how we provide docs at that level) falling on the Product Education team. Content beyond reference (to include How to, Tutorials, and Explanation will be the primary focus of net new content by Product Education. I totally expect that other teams will contribute to that as well so it's not intended to be exclusive. Ideally, all this content will live in close proximity to the docs (for example via proper help/content portal) long term, but in the interim will be organized within what we are calling "docs" today.
5. When we create docs of any kind, we will focus on providing high-value content specific to our product. In addition, we should make sure we are not documenting technologies that are already well documented unless we intend (and accept the cost) of both maintaining those docs and potentially misguiding our customers with content that can't be updated as quickly as that made available by the technology provider. For example, AWS docs will always be the most up-to-date, and it would be orthogonal to document elements of their interface as part of our instruction (which there are examples of in our docs). This has been a topic discussed and is aligned with guidance from our Delivery product management and engineering teams. If we decide that we do intend/need to document those technologies more fully, we should also be very transparent about what support we can and will provide. In my opinion, the best approach is to provide information that helps regardless of the technology provider. For example, if you utilize any cloud services, what are some shared best practices with regard to deployment? This is different than providing all the steps to perform an action, which can be really problematic based on the reasons mentioned above.

> This tool is in the docs and isn't useful in it's current state, to the extent that we have been calling out that working with a CE is going to be a better route. I'm going to keep the link to the. resource estimator and the suggested change above of adding in CE language, but I wanted to make sure this is elevated again that we're saying here that our official docs aren't sufficient without CE involvement.

This is a known issue, my understanding is there is active (or will be active) development in this area in this quarter, though it is being prioritized based on other requests and deliverables. @anna-mmikhova and @caugustus-sourcegraph are most likely the points of contact for current state. Though cc @briemcnally for posterity.

> `Do` you have a suggestion on how we can point to specifics needed here based on how the docs are now without writing something that should instead be in the docs? I'm coming up blank on that, which is why I'm pushing instead for the official documentation to be updated.  

The best approach in a situation like this is to set expectations before someone clicks the link, and also on the page. If additional knowledge is needed (like you mention it makes sense to you, let's add that as a stop-gap). The likelihood that we update this specific area soon may be low given all the points above.

> I'm not sure. That doc is great, and looks like something that would be perfect for a deployment guide that lives in the docs. @jwayman what do you think about a way to integrate guides like this into the docs, because this kind of step by step stuff is VERY valuable. This reference shouldn't get into details step by step for it's purpose, but this kind of step by step thing is exactly what I'd expect to find in a "guides" section of official documentation for any product I was wanting to use.

I am totally behind step-by-step docs. Many of the docs we have today miss this. My concern would where should we draw the line on what we document.

The part that concerns me here is how do we know these steps (which reference specific interface elements and steps with regard to k8s) are accurate and up-to-date? I think the logic with this doc is that it is intended as internal support/documentation (which may live in the handbok). 

Where we have fallen in regard to that question in recent conversations is that we want to focus on the most straightforward path for self-service in an effort to avoid any potential pitfalls for users going that route. In looking at best-in-class examples approaches [like this](https://docs.github.com/en/enterprise-server@3.1/actions/deployment/deploying-to-your-cloud-provider/deploying-to-google-kubernetes-engine), which focus more on step-by-step at the command line are probably a better approach. 

We can revisit this, but I think we should be very clear with regard to the risk we accept and how we plan to support customers if our content is not updated. CCing some members of the engineering team for some thoughts here as well (@michaellzc @kevinwojo @caugustus-sourcegraph).

> I do have to ask though: Why do we have that detailed setup guide for GCP in the handbook, but then in our docs we have [cloud installation](https://docs.sourcegraph.com/admin/deploy/kubernetes#cloud-installation) linking to [GKE's official docs](https://cloud.google.com/kubernetes-engine/docs/quickstart)

I think there should be enough context to some degree, but just to clarify there is actually another existing [Google Doc for K*s and GKE](https://docs.google.com/document/d/1NKozoBSpAj9EDklXeZYItq6wRYep7lvGfpxYd-jRJzw/edit) as well. Based on a conversation we had in the User Education channel a couple weeks back, I believe the plan is to migrate from the Google Doc to the docs. That would be a relatively easy activity.

As to why we have the Doc in the handbook, I think that's a case of we have some internal documentation - which the handbook has been the source of truth for those things - that page is ~6 months old though, so there could have been a change in direction.

As to the docs linking to GKE, this is ideally where we want to be to ensure users have the most accurate documentation related to the technology path they have chosen. Again, to date, we have not intended for the docs to re-document things documented by those technology providers, and to ensure the most accurate content and avoid pitfalls linking to those docs is [generally a best practice](https://docs.github.com/en/enterprise-server@3.1/actions/deployment/deploying-to-your-cloud-provider/deploying-to-google-kubernetes-engine#creating-a-gke-cluster).

> this is the kind of thing that I personally feel we need to correct. This handout should be able to reference and walk through all of our official documentation. What I'm hearing in some feedback in the PR right now is that the docs are not sufficient in terms of the depth of explanation, and/or do not explain at all and we link to an external doc (like in GKE's instance).

This isn't a surprise. Per the first point, docs have had no centralized custodian/champion/curator for many years, there is a tremendous amount of debt. Add to that, of the efforts that have been made there has been no focus on K8s deployment without Helm. In fact, we strategically were moving away from that, and I'd expect this area to have some of the biggest gaps because of that. If we want to focus efforts here, I think we should make sure to get everyone on the same page, as these gaps also exist in many other areas as well.

> I stand by that this doc should reference Sourcegraph official documentation. If we have better documentation in Google Drive, in the Handbook, in a blog post, etc. why isn't it instead migrated to the documentation so we can link to a single source of truth?

I hope I addressed the why to your question here. I think som broader discussions about where we focus our efforts and prioritize (since we can't do everything) and there is a long list of colliding priorities for new and updated deployment/setup content in this quarter alone. Not to mention, failing IA that prevents user getting to content that actually would help them.

**Comment by caugustus-sourcegraph** [comment]
@IAmHughes @jwayman - My initial response to looking at this PR without reading Jeff's comment was: "Wow, this would sure be nice to include in the official docs by default".

k8s without helm isn't our default recommendation anymore, but it's still an officially supported option (and will be for quite some time). Jeff raised a valid concern about maintenance and risk. I'm less worried about that for a doc like this, which will be actively used by CE's. Documentation is owned by everyone, so if these are resources you would find valuable, I lean towards incorporating it.

File: collateral/deployment/k8s-without-helm-reference.md
Comment (emchap) [reply collateral/deployment/k8s-without-helm-reference.md:33]:
```suggestion
Sourcegraph has a list of configurations available and recommended that need to be made to your manifest file to properly install Sourcegraph into your cluster. This includes how to customize the deployment using Kustomize—if you prefer to use Helm, please see the [Helm install instructions](https://docs.sourcegraph.com/admin/deploy/kubernetes/helm), instead. 

Please review the full configuration guide prior to proceeding: [Configure Sourcegraph with Kubernetes](https://docs.sourcegraph.com/admin/deploy/kubernetes/configure). Pay particular attention to configuring a storage class for your cloud provider, configuring network access, and configuring external databases if you choose to use them.

Other sections of the configuration page should be reviewed, but are not required (for example, an external Jaeger instance is optional).
```

**Comment by jwayman** [comment]
@caugustus-sourcegraph +1 we could certainly add this to the docs. @IAmHughes and I had chatted about this, but I think the desire was not to have it there, but as a resource for CEs. 

**I'm totally down for putting outlines/checklists like this into the docs. In fact a sort into/explanatory/checklist to pair with the reference material would be great.** 

My comments were mostly focused on prioritizing effort to focus on documenting K8s w/out Helm and efforts to make that path self-service. I don't believe that would be on my (or other) teams' radar any time soon - unless there has been a shift I'm not aware of.

**Review by emchap — APPROVED** [review]
Discussed with Thomas in DM. I think this is fine to push as-is; the issues with the content can be modified as needed depending on customer feedback, and the other conversations, though interesting, are ideological in nature.

**Pull request merged**
